Add standard Slack messages and colors

ddgenome opened the pull request:
Add standard formatted Slack messages for success, error, and
warning.  Add standard colors for each case.

tanya-coding commented on src/SlackMessages.ts:
```diff
@@ -40,9 +57,9 @@ export function url(fullUrl: string, label?: string) {
  */
 export function user(userId: string, userName?: string) {
-    if (!emptyString(userId) && !emptyString(userName)) {
+    if (userId && userName) {
```
Same here, we don't want to try to convert empty strings into `<|>` or `<>`. Same applies to channel, bold (** will render as 2 starts which is not correct), italic, etc.

tanya-coding commented on src/SlackMessages.ts:
```diff
@@ -98,61 +115,37 @@ export function render(message: SlackMessage, pretty: boolean = false): string {
 /** Render emoji by name */
 export function emoji(name: string) {
-    return `:${name}:`;
+    return (name) ? `:${name}:` : "";
```
We should add emptyString check here

tanya-coding commented on src/SlackMessages.ts:
```diff
@@ -98,61 +115,37 @@ export function render(message: SlackMessage, pretty: boolean = false): string {
-    } else {
-        return "";
-    }
+    return (text) ? `*${text}*` : "";
```
I think we still need emptyString check here

tanya-coding commented on src/SlackMessages.ts:
```diff
@@ -98,61 +115,37 @@ export function render(message: SlackMessage, pretty: boolean = false): string {
-    } else {
-        return "";
-    }
+    return (text) ? `_${text}_` : "";
```
I think we still need emptyString check here

tanya-coding commented on src/SlackMessages.ts:
```diff
@@ -98,61 +115,37 @@ export function render(message: SlackMessage, pretty: boolean = false): string {
-    } else {
-        return "";
-    }
+    return (text) ? `~${text}~` : "";
```
I think we still need emptyString check here

tanya-coding commented on src/SlackMessages.ts:
```diff
@@ -98,61 +115,37 @@ export function render(message: SlackMessage, pretty: boolean = false): string {
-    } else {
-        return "";
-    }
+    return (text) ? "`" + text + "`" : "";
```
I think we still need emptyString check here

tanya-coding commented on src/SlackMessages.ts:
````diff
@@ -98,61 +115,37 @@ export function render(message: SlackMessage, pretty: boolean = false): string {
-    } else {
-        return "";
-    }
+    return (text) ? "```" + text + "```" : "";
````
I think we still need emptyString check here

tanya-coding commented on src/SlackMessages.ts:
```diff
@@ -98,61 +115,37 @@ export function render(message: SlackMessage, pretty: boolean = false): string {
-    } else {
-        return "";
-    }
+    return (item) ? `• ${item}` : "";
```
I think we still need emptyString check here

tanya-coding commented on src/SlackMessages.ts:
```diff
@@ -237,7 +230,7 @@ export class ValidationError extends Error {
 /** Construct Slack button that will execute provided rug instruction. */
 export function rugButtonFrom(action: ButtonSpec, command: IdentifiableInstruction): Action {
-    if (emptyString(command.id)) {
+    if (!command.id) {
```
I don't think we should accept empty string as action ID

tanya-coding commented on src/SlackMessages.ts:
```diff
@@ -253,3 +246,97 @@ export function rugButtonFrom(action: ButtonSpec, command: IdentifiableInstructi
     return button;
 }
+
+interface StandardAttachmentDefaults {
```
Should we move these into a separate file in order to keep slack-messages more generic?

ddgenome replied:
I'm happy to move it to a different file. I just though because these were Slack messages, they belonged in SlackMessages.

tanya-coding replied:
There is RugMessages.ts where something similar is already being attempted - maybe just replace/enhance what is done there with new standard messages?

tanya-coding commented:
Added some comments. I think removing emptyString check in quite a few functions will have negative effect

ddgenome commented:
@tanya-coding the empty string is false in TypeScript/JavaScript (https://javascriptweblog.wordpress.com/2011/02/07/truth-equality-and-javascript/), so you will get the proper result if the argument passed in is `null`, `undefined`, or the empty string.

Essentially, the `emptyString` test is a much more complicated version of what JavaScript does by default.

ddgenome commented:
I've added SlackMessages test cases for empty string input everywhere there were tests cases for `null` and/or `undefined`. They all pass.

tanya-coding commented:
LGTM 👍

ddgenome commented:
@tanya-coding I've moved the standard message stuff to RugMessages, added versions that return ResponseMessages, and deprecated renderError and renderSuccess. I deprecated rather than updated the latter two because I felt the changes were sufficiently different that I didn't want to change the message format out from people.
